include/logging/log_core.h

```diff
@@ -185,6 +185,49 @@ extern "C" {
 		}							 \
 	} while (false)
 
+static inline log_arg_t __do_strdup(u32_t msk, u32_t idx, log_arg_t param)
+{
+	if (msk & (1<<idx)) {
+		const char *str = (const char *)param;
+		char *log_strdup(const char *str);
+
+		param = (log_arg_t)log_strdup(str);
+	}
+	return param;
+}
+
+#define __LOG_INTERNAL_VA(is_user_context, _src_level, _format, _valist, _argnum, _strdup) \
+	do {								 \
+		if (is_user_context) {					 \
+			log_generic_from_user(_src_level, _format, _valist);\
+		} else if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {		 \
+			log_generic(_src_level, _format, _valist, _strdup);\
+		} else {						 \
+			if (_argnum == 0)				\
+				_LOG_INTERNAL_0(_src_level, _format); \
+			else {						\
+			u32_t mask = _strdup?z_log_get_s_mask(_format, _argnum):0;\
+										\
+			if (_argnum == 1)				\
+				_LOG_INTERNAL_1(_src_level, _format, \
```

Comment:
dedicated logging functions for 0-3 arguments were created for performance. Here, we are giving up on that so it would be better to use `log_generic` for all cases.

Reply:
Why shall we give up? The path for 0-3 args will be shorter. oPlatLog is a log sink for 600 OT files

Reply:
is `_argnum` known at compile time? if not that can impact flash size. Also reason for having dedicated functions was to not used va_list arguments because it was significant addition to logging time in the optimized scenario (~30-40% longer logging time). Here you already have va_list so i don't think that overall performance gain is significant. Maybe you can measure that. Anyway, it's up to you to check and decide flash size penalty vs performance.

Reply:
Actually, those macros are to be used only 4 times in zephyr code, inside single function otPlatLog. Compilator will do further optimization, because the macro will differ only with log level. Size increase will be very little. Because all logs from OpenThread will go via this function, it's speed is much more important than it's size.
But I have a question about further speed performance:
We could avoid %s searching for none-zero arguments number, if we know, that format string contains %s or not; format string is known at compile time! Do you know some legal method to check at compile time, that "string" generated by a macro, has some specific value?

Reply:
right, if that's the case then performance has more value.

> Do you know some legal method to check at compile time, that "string" generated by a macro, has some specific value?

unfortunately, not.